File: Core/src/main/java/com/github/ezauton/core/action/PurePursuitAction.java
package com.github.ezauton.core.action;

import com.github.ezauton.core.localization.TranslationalLocationEstimator;
import com.github.ezauton.core.pathplanning.PathSegment;
import com.github.ezauton.core.pathplanning.Path;
import com.github.ezauton.core.pathplanning.purepursuit.Lookahead;
import com.github.ezauton.core.pathplanning.purepursuit.PurePursuitMovementStrategy;
import com.github.ezauton.core.robot.subsystems.TranslationalLocationDriveable;
import com.github.ezauton.core.trajectory.geometry.ImmutableVector;

import java.util.concurrent.TimeUnit;

/**
 * A Pure Pursuit action which can be used in simulation or as a WPILib Command
 */
public final class PurePursuitAction extends PeriodicAction  // TODO: Rename to PPAction
{
    private final PurePursuitMovementStrategy purePursuitMovementStrategy;
    private final TranslationalLocationEstimator translationalLocationEstimator;
    private final Lookahead lookahead;
    private final TranslationalLocationDriveable translationalLocationDriveable;
    private double speedUsed;
    private double absoluteDistanceUsed;

    /**
     * Create a PP Command
     *
     * @param period                         How often to update estimated position, robot control, etc
     * @param timeUnit                       The timeunit that period is in
     * @param purePursuitMovementStrategy    Our movement strategy.
     * @param translationalLocationEstimator An object that knows where we are on a 2D plane
     * @param lookahead                      An instance of {@link Lookahead} that can tell us how far along the path to look ahead
     * @param translationalLocationDriveable The drivetrain of the robot
     */
    public PurePursuitAction(long period, TimeUnit timeUnit, PurePursuitMovementStrategy purePursuitMovementStrategy, TranslationalLocationEstimator translationalLocationEstimator, Lookahead lookahead, TranslationalLocationDriveable translationalLocationDriveable) {
        super(period, timeUnit);
        this.purePursuitMovementStrategy = purePursuitMovementStrategy;
        this.translationalLocationEstimator = translationalLocationEstimator;
        this.lookahead = lookahead;
        this.translationalLocationDriveable = translationalLocationDriveable;
    }

    @Override
    public void execute() {
        // Find out where to drive to
        ImmutableVector loc = translationalLocationEstimator.estimateLocation();
        ImmutableVector goalPoint = purePursuitMovementStrategy.update(loc, lookahead.getLookahead());

        Path path = purePursuitMovementStrategy.getPath();
        PathSegment current = path.getCurrent();
        ImmutableVector closestPoint = current.getClosestPoint(loc);
        absoluteDistanceUsed = current.getAbsoluteDistance(closestPoint);
        speedUsed = current.getSpeed(absoluteDistanceUsed);
        translationalLocationDriveable.driveTowardTransLoc(speedUsed, goalPoint);
    }

    /**
     * @return The most reset speed used by Pure Pursuit
     */
    public double getSpeedUsed() {
        return speedUsed;
    }

    /**
     * @return The most reset absolute distance used by Pure Pursuit
     */
    public double getAbsoluteDistanceUsed() {
        return absoluteDistanceUsed;
    }

    @Override
    public boolean isFinished() {
        if (purePursuitMovementStrategy.isFinished()) {
            translationalLocationDriveable.driveSpeed(0);
            return true;
        }
        return false;
    }
}

File: Core/src/main/kotlin/com/github/ezauton/core/action/PurePursuitAction.kt
package com.github.ezauton.core.action

import com.github.ezauton.conversion.*
import com.github.ezauton.core.localization.TransLocEst
import com.github.ezauton.core.pathplanning.PathProgressor
import com.github.ezauton.core.pathplanning.Trajectory
import com.github.ezauton.core.pathplanning.purepursuit.Lookahead
import com.github.ezauton.core.pathplanning.purepursuit.PurePursuitMovementStrategy
import com.github.ezauton.core.pathplanning.purepursuit.Update
import com.github.ezauton.core.record.Data
import com.github.ezauton.core.record.RecordingContext
import com.github.ezauton.core.robot.subsystems.TransLocDrivable
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.consumeAsFlow


typealias Speed = (distance: Distance) -> LinearVelocity

/**
 * A Pure Pursuit action which can be used in simulation or as a WPILib Command
 *
 * @param period How often to update estimated position, robot control, etc
 * @param translationalLocationEstimator An object that knows where we are on a 2D plane
 * @param lookahead An instance of [Lookahead] that can tell us how far along the path to look ahead
 * @param translationalLocationDrivable The drivetrain of the robot
 */
fun purePursuit(
  period: Period,
  trajectory: Trajectory,
  translationalLocationEstimator: TransLocEst,
  translationalLocationDrivable: TransLocDrivable,
  lookahead: Lookahead,
  stopDistance: Distance = 0.01.m,
) = action {

  val recordingContext = coroutineContext[RecordingContext]
  val dataChannel = if (recordingContext == null) null else Channel<Data.PurePursuit>()

  recordingContext?.recording?.receiveFlow(dataChannel!!.consumeAsFlow())

  val progressor = PathProgressor(trajectory.path)
  val speedFunction = trajectory.speed
  val ppMoveStrat = PurePursuitMovementStrategy(progressor, stopDistance, dataChannel)

  coroutineScope {

    periodic(period) { loop ->

      val currentLocation = translationalLocationEstimator.estimateLocation()

      when (val update = ppMoveStrat.update(currentLocation, lookahead.lookahead)) {
        is Update.Finished -> {
          translationalLocationDrivable.driveSpeed(0.0.mps)
          loop.stop()
        }
        is Update.Result -> {
          val speedUsed = speedFunction(update.on.distance)
          translationalLocationDrivable.driveTowardTransLoc(speedUsed, update.goal)
        }
      }
    }

  }
  dataChannel?.close()
  return@action
}

fun <T> T.purePursuit(period: Period, trajectory: Trajectory, lookahead: Lookahead, stopDistance: Distance = 0.01.m) where T : TransLocEst, T : TransLocDrivable =
  purePursuit(period, trajectory, this, this, lookahead, stopDistance)

fun <T> T.purePursuit(period: Time, trajectory: Trajectory, lookahead: Lookahead, stopDistance: Distance = 0.01.m) where T : TransLocEst, T : TransLocDrivable =
  purePursuit(Period(period), trajectory, this, this, lookahead, stopDistance)
